Content.Shared/Body/Systems/SharedBloodstreamSystem.cs

```diff
@@ -289,25 +294,53 @@ private void OnRejuvenate(Entity<BloodstreamComponent> ent, ref RejuvenateEvent
     {
         TryModifyBleedAmount(ent.AsNullable(), -ent.Comp.BleedAmount);
 
-        if (SolutionContainer.ResolveSolution(ent.Owner, ent.Comp.BloodSolutionName, ref ent.Comp.BloodSolution, out var bloodSolution))
-            TryModifyBloodLevel(ent.AsNullable(), bloodSolution.AvailableVolume);
+        if (SolutionContainer.ResolveSolution(ent.Owner, ent.Comp.BloodSolutionName, ref ent.Comp.BloodSolution))
+        {
+            SolutionContainer.RemoveAllSolution(ent.Comp.BloodSolution.Value);
+            TryModifyBloodLevel(ent.AsNullable(), ent.Comp.BloodReferenceVolume);
+        }
+    }
+
+    private void OnMetabolismExclusion(Entity<BloodstreamComponent> ent, ref MetabolismExclusionEvent args)
+    {
+        // We explicitly want the exactly blood type that way consuming others blood makes us sick
+        var referenceBlood = new ReagentId(ent.Comp.BloodReagent, GetEntityBloodData(ent.Owner));
 
-        if (SolutionContainer.ResolveSolution(ent.Owner, ent.Comp.ChemicalSolutionName, ref ent.Comp.ChemicalSolution))
-            SolutionContainer.RemoveAllSolution(ent.Comp.ChemicalSolution.Value);
+        // Copy to list again to avoid modifying during enumeration!
+        foreach (var reagentQuantity in args.ReagentList.ToList())
+        {
+            if (reagentQuantity.Reagent != referenceBlood)
+                continue;
+
+            args.ReagentList.Remove(reagentQuantity);
+        }
     }
 
     /// <summary>
     /// Returns the current blood level as a percentage (between 0 and 1).
     /// </summary>
-    public float GetBloodLevelPercentage(Entity<BloodstreamComponent?> ent)
+    public float GetBloodLevel(Entity<BloodstreamComponent?> entity)
     {
-        if (!Resolve(ent, ref ent.Comp)
-            || !SolutionContainer.ResolveSolution(ent.Owner, ent.Comp.BloodSolutionName, ref ent.Comp.BloodSolution, out var bloodSolution))
+        if (!Resolve(entity, ref entity.Comp)
+            || !SolutionContainer.ResolveSolution(entity.Owner, entity.Comp.BloodSolutionName, ref entity.Comp.BloodSolution, out var bloodSolution)
+            || entity.Comp.BloodReferenceVolume == 0)
         {
             return 0.0f;
         }
 
-        return bloodSolution.FillFraction;
+        var totalBloodLevel = 0.0f;
+
+        for (var i = bloodSolution.Contents.Count - 1; i >= 0; i--)
+        {
+            // Unlike with the metabolism exclusion, we treat all blood as blood, so long as it's "close enough"
+            var (reagentId, quantity) = bloodSolution.Contents[i];
+            if (reagentId.Prototype == entity.Comp.BloodReagent)
+            {
+                totalBloodLevel += quantity.Float();
+            }
```

Comment:
so foreign blood in your bloodstream counts as blood for the purposes of blood level but will be metabolized and won't be affected by methods that modify the blood level explicitly...?